File: src/routing.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <array>
#include <cmath>
#include <iomanip>
#include <stdexcept>

struct TimeStep {
    double dt_hours;
    double q_in_m3s;
};

struct ReservoirState {
    double storage_m3 = 0.0;
};

struct ReservoirStepResult {
    double storage_new_m3;
    double q_out_avg_m3s;
    double q_out_end_m3s;
};

std::vector<TimeStep> loadInflowFile(const std::string& filename) {
    std::ifstream in(filename);
    if (!in.is_open()) {
        throw std::runtime_error("Could not open file: " + filename);
    }

    std::vector<TimeStep> series;
    std::string line;
    size_t line_no = 0;

    while (std::getline(in, line)) {
        ++line_no;

        size_t hash_pos = line.find('#');
        if (hash_pos != std::string::npos) {
            line.erase(hash_pos);
        }

        if (line.find_first_not_of(" \t\r\n") == std::string::npos) {
            continue;
        }

        std::istringstream iss(line);
        TimeStep step{};

        if (!(iss >> step.dt_hours >> step.q_in_m3s)) {
            if (line_no == 1) {
                continue;
            }
            throw std::runtime_error(
                "Parse error in " + filename + " at line " + std::to_string(line_no)
            );
        }

        if (step.dt_hours <= 0.0) {
            throw std::runtime_error(
                "dt_hours must be positive in " + filename + " at line " + std::to_string(line_no)
            );
        }

        if (step.q_in_m3s < 0.0) {
            throw std::runtime_error(
                "q_in_m3s must be non-negative in " + filename + " at line " + std::to_string(line_no)
            );
        }

        series.push_back(step);
    }

    if (series.empty()) {
        throw std::runtime_error("No valid data rows in file: " + filename);
    }

    return series;
}

class ThreeReservoirCascade {
public:
    explicit ThreeReservoirCascade(double k_total_hours) {
        k_total_seconds_ = k_total_hours * 3600.0;
        k_res_seconds_ = k_total_seconds_ / 3.0;
    }

    void setInitialStorage(double total_storage_m3) {
        double share = total_storage_m3 / 3.0;
        for (size_t i = 0; i < reservoirs_.size(); ++i) {
            reservoirs_[i].storage_m3 = share;
        }
    }

    double route(double q_in_m3s, double dt_seconds) {
        double q_for_next = q_in_m3s;

        for (size_t i = 0; i < reservoirs_.size(); ++i) {
            ReservoirStepResult result =
                routeOneReservoir(reservoirs_[i].storage_m3, q_for_next, k_res_seconds_, dt_seconds);

            reservoirs_[i].storage_m3 = result.storage_new_m3;
            q_for_next = result.q_out_avg_m3s;
        }

        return q_for_next; // average outlet discharge over the step
    }

    double totalStorageM3() const {
        double sum = 0.0;
        for (size_t i = 0; i < reservoirs_.size(); ++i) {
            sum += reservoirs_[i].storage_m3;
        }
        return sum;
    }

    double reservoirKHours() const {
        return k_res_seconds_ / 3600.0;
    }

private:
    ReservoirStepResult routeOneReservoir(
        double storage_old_m3,
        double q_in_m3s,
        double k_seconds,
        double dt_seconds)
    {
        double factor = std::exp(-dt_seconds / k_seconds);

        double storage_new_m3 =
            storage_old_m3 * factor
            + k_seconds * (1.0 - factor) * q_in_m3s;

        double volume_out_m3 =
            storage_old_m3 + q_in_m3s * dt_seconds - storage_new_m3;

        double q_out_avg_m3s = volume_out_m3 / dt_seconds;
        double q_out_end_m3s = storage_new_m3 / k_seconds;

        if (q_out_avg_m3s < 0.0) q_out_avg_m3s = 0.0;
        if (q_out_end_m3s < 0.0) q_out_end_m3s = 0.0;

        return {storage_new_m3, q_out_avg_m3s, q_out_end_m3s};
    }

    std::array<ReservoirState, 3> reservoirs_;
    double k_total_seconds_ = 0.0;
    double k_res_seconds_ = 0.0;
};

int main() {
    try {
        const std::string inflow_filename = "inflow.txt";
        const std::string out_filename = "res.txt";

        const double k_total_hours = 5.0;
        const double initial_storage_m3 = 100000.0;

        std::vector<TimeStep> series = loadInflowFile(inflow_filename);

        ThreeReservoirCascade channel(k_total_hours);
        channel.setInitialStorage(initial_storage_m3);

        std::ofstream out(out_filename);
        if (!out.is_open()) {
            throw std::runtime_error("Could not open output file: " + out_filename);
        }

        out << std::fixed << std::setprecision(6);
        out << "step\tdt_hours\tQin_m3s\tQout_m3s\tstorage_m3\n";

        double total_in_m3 = 0.0;
        double total_out_m3 = 0.0;

        for (size_t i = 0; i < series.size(); ++i) {
            
            double dt_seconds = series[i].dt_hours * 3600.0;
            double q_in = series[i].q_in_m3s;
            double q_out = channel.route(q_in, dt_seconds);
            double storage_m3 = channel.totalStorageM3();

            total_in_m3 += q_in * dt_seconds;
            total_out_m3 += q_out * dt_seconds;

            out << (i + 1) << "\t"
                << series[i].dt_hours << "\t"
                << q_in << "\t"
                << q_out << "\t"
                << storage_m3 << "\n";
        }

        out.close();

        double final_storage_m3 = channel.totalStorageM3();
        double water_balance_m3 = initial_storage_m3 + total_in_m3 - total_out_m3 - final_storage_m3;

        std::cout << std::fixed << std::setprecision(3);
        std::cout << "3-reservoir cascade routing completed\n";
        std::cout << "K_total_hours      = " << k_total_hours << "\n";
        std::cout << "K_reservoir_hours  = " << channel.reservoirKHours() << "\n";
        std::cout << "Output written to  = " << out_filename << "\n";
        std::cout << "Initial storage    = " << initial_storage_m3 << " m3\n";
        std::cout << "Total inflow       = " << total_in_m3 << " m3\n";
        std::cout << "Total outflow      = " << total_out_m3 << " m3\n";
        std::cout << "Final storage      = " << final_storage_m3 << " m3\n";
        std::cout << "Water balance      = " << water_balance_m3 << " m3\n";

        return 0;
    }
    catch (const std::exception& e) {
        std::cerr << "ERROR: " << e.what() << "\n";
        return 1;
    }
}

File: src/routing.rs
use crate::error::{HerssError, Result};
use crate::{MAX_NR_CASCADED_RESERVOIRS, MAX_TRAVELTIME_HOURS};

#[derive(Debug, Clone, Copy)]
struct ReservoirStepResult {
    storage_new_m3: f64,
    q_out_avg_m3s: f64,
}

#[derive(Debug, Clone)]
pub(crate) struct CascadedReservoirs {
    storage_m3: Vec<f64>,
    k_res_seconds: f64,
}

impl CascadedReservoirs {
    pub fn new(k_traveltime_hours: f64, num_reservoirs: usize) -> Result<Self> {
        if !(0.0..=MAX_TRAVELTIME_HOURS).contains(&k_traveltime_hours) {
            return Err(HerssError::new(format!(
                "Travel time must be between 0 and {MAX_TRAVELTIME_HOURS} hours"
            )));
        }
        if !(1..=MAX_NR_CASCADED_RESERVOIRS).contains(&num_reservoirs) {
            return Err(HerssError::new(format!(
                "Number of cascaded reservoirs must be between 1 and {MAX_NR_CASCADED_RESERVOIRS}"
            )));
        }

        Ok(Self {
            storage_m3: vec![0.0; num_reservoirs],
            k_res_seconds: k_traveltime_hours * 3600.0 / num_reservoirs as f64,
        })
    }

    pub fn set_initial_storage(&mut self, initial_storage_mm3: &[f64]) -> Result<()> {
        if initial_storage_mm3.len() != self.storage_m3.len() {
            return Err(HerssError::new(
                "Initial storage size does not match the number of reservoirs",
            ));
        }
        for (idx, value) in initial_storage_mm3.iter().copied().enumerate() {
            if value < 0.0 {
                return Err(HerssError::new("Initial storage cannot be negative"));
            }
            self.storage_m3[idx] = value * 1_000_000.0;
        }
        Ok(())
    }

    pub fn route(&mut self, q_in_m3s: f64, dt_seconds: f64) -> f64 {
        let mut q_for_next = q_in_m3s;
        for storage in &mut self.storage_m3 {
            let result = route_one_reservoir(*storage, q_for_next, self.k_res_seconds, dt_seconds);
            *storage = result.storage_new_m3;
            q_for_next = result.q_out_avg_m3s;
        }
        q_for_next
    }

    pub fn total_storage_m3(&self) -> f64 {
        self.storage_m3.iter().sum()
    }

    pub fn storage_mm3(&self, idx: usize) -> Result<f64> {
        self.storage_m3
            .get(idx)
            .map(|value| value / 1_000_000.0)
            .ok_or_else(|| HerssError::new("Index out of range in getStorageMm3"))
    }
}

fn route_one_reservoir(
    storage_old_m3: f64,
    q_in_m3s: f64,
    k_seconds: f64,
    dt_seconds: f64,
) -> ReservoirStepResult {
    if k_seconds <= 0.0 {
        let q_out_avg_m3s = (storage_old_m3 + q_in_m3s * dt_seconds) / dt_seconds;
        return ReservoirStepResult {
            storage_new_m3: 0.0,
            q_out_avg_m3s: q_out_avg_m3s.max(0.0),
        };
    }

    let factor = (-dt_seconds / k_seconds).exp();
    let storage_new_m3 = storage_old_m3 * factor + k_seconds * (1.0 - factor) * q_in_m3s;
    let volume_out_m3 = storage_old_m3 + q_in_m3s * dt_seconds - storage_new_m3;
    ReservoirStepResult {
        storage_new_m3,
        q_out_avg_m3s: (volume_out_m3 / dt_seconds).max(0.0),
    }
}

#[cfg(test)]
mod tests {
    use super::CascadedReservoirs;

    #[test]
    fn one_hour_routing_matches_formula() {
        let mut route = CascadedReservoirs::new(1.0, 1).unwrap();
        route.set_initial_storage(&[0.0]).unwrap();
        let q = route.route(2.0, 3600.0);
        let factor = (-1.0f64).exp();
        let expected_storage = 3600.0 * (1.0 - factor) * 2.0;
        let expected_q = ((2.0 * 3600.0) - expected_storage) / 3600.0;
        assert!((q - expected_q).abs() < 1e-12);
    }
}
